[WIP] Better `sim.lattice` (#1381)
BLAST-ImpactX/impactx: docs/source/install/upgrade.rst

   phase_shift = rf.phase          # once, before tracking

   def hook_before_element(sim):
       element = sim.tracking_element
       if type(element) is elements.RFCavity:
           element.phase = optimize(sim.beam.ref, element) + phase_shift

The same applies to the free functions ``impactx.reverse(element)`` and
``impactx.push(pc, element)``, which now act on the element passed in.

Changing the lattice during tracking
""""""""""""""""""""""""""""""""""""

Adding or removing elements from a tracking hook raises, because tracking is walking the
sequence at that moment. Changing an element's own parameters from a hook is unaffected
and remains the intended way to retune a lattice mid-run.

``sim.lattice`` needs its simulation
""""""""""""""""""""""""""""""""""""

A lattice reached through ``sim.lattice`` belongs to that simulation and does not keep it
alive. Using the lattice after the simulation is gone raises, where it used to keep
working:

.. code-block:: python

   def build():
       sim = ImpactX()
       sim.lattice.extend(cell)
       return sim.lattice        # the simulation ends here

   lattice = build()
   len(lattice)                  # raises

> Return the simulation, or keep a reference to it for as long as you use its lattice.

Author: ```suggestion
Return the simulation, or keep a reference to it for as long as you use its lattice.
Alternatively, keep the lattice alive across simulations by storing it in a pure Python list and call `sim.lattice.extend` directly before tracking.
```

Selections describe the lattice they were taken from

> """"""""""""""""""""""""""""""""""""""""""""""""""""""

Author: ```suggestion
Selections describe the lattice they were taken from
""""""""""""""""""""""""""""""""""""""""""""""""""""
```

reporting the wrong elements:

.. code-block:: python

   quads = sim.lattice.select(kind="Quad")
   sim.lattice.append(elements.Drift(ds=0.1))
   quads[0]                      # raises; select() again

Changing an element's own parameters moves nothing and leaves a selection usable.

``select(kind=...)`` matches the element kind
"""""""""""""""""""""""""""""""""""""""""""""

An element written as a Python subclass is still of its element kind, so
``select(kind="Drift")`` and ``select(kind=elements.Drift)`` now match it. The name of the
subclass matches as well.

``replace_each()`` needs a template it can copy
"""""""""""""""""""""""""""""""""""""""""""""""

The template is copied into each selected position, so a Python subclass must define
``copy()``. Passing one that does not raises, where it previously inserted plain
base-class elements.

``finalize()`` always empties the lattice
"""""""""""""""""""""""""""""""""""""""""

``sim.finalize()`` releases the elements whether or not ``init_grids()`` was ever called.
They have been finalized by then, so keeping them would leave elements that are done with

> -- a beam monitor with its output closed, say -- still in the lattice.

Author: ```suggestion
, e.g., a beam monitor with its output closed, still in the lattice.
```

New in this release
"""""""""""""""""""

- ``element.copy()`` on every element type, for a new element with the same configuration.
- The array-valued element parameters -- Fourier coefficients, multipole coefficients and

>   polygon vertices -- can be set after construction, via their properties or the paired

Author: ```suggestion
- The array-valued element parameters (Fourier coefficients, multipole coefficients and
  polygon vertices) can be set after construction, via their properties or the paired
```
